src/Donations/Repositories/DonationRepository.php

```diff
@@ -0,0 +1,519 @@
+                        'meta_value' => $metaValue,
+                    ]);
+            }
+        } catch (Exception $exception) {
+            DB::query('ROLLBACK');
+
+            Log::error('Failed creating a donation');
+
+            throw new $exception('Failed creating a donation');
+        }
+
+        DB::query('COMMIT');
+
+        $donation = $this->getById($donationId);
+
+        Hooks::dispatch('give_donation_created', $donation);
+
+        return $donation;
+    }
+
+    /**
+     * @unreleased
+     *
+     * @param  Donation  $donation
+     *
+     * @return Donation
+     * @throws Exception|InvalidArgumentException
+     */
+    public function update(Donation $donation)
+    {
+        $this->validateDonation($donation);
+
+        Hooks::dispatch('give_donation_updating', $donation);
+
+        $date = Temporal::getCurrentFormattedDateForDatabase();
+
+        DB::query('START TRANSACTION');
+
+        try {
+            DB::table('posts')
+                ->where('ID', $donation->id)
+                ->update([
+                    'post_modified' => $date,
+                    'post_modified_gmt' => get_gmt_from_date($date),
+                    'post_status' => $donation->status->getValue(),
+                    'post_type' => 'give_payment',
+                    'post_parent' => isset($donation->parentId) ? $donation->parentId : 0
+                ]);
+
+            foreach ($this->getCoreDonationMetaForDatabase($donation) as $metaKey => $metaValue) {
+                DB::table('give_donationmeta')
+                    ->where('donation_id', $donation->id)
+                    ->where('meta_key', $metaKey)
+                    ->update([
+                        'meta_value' => $metaValue,
+                    ]);
+            }
+        } catch (Exception $exception) {
+            DB::query('ROLLBACK');
+
+            Log::error('Failed updating a donation');
+
+            throw new $exception('Failed updating a donation');
+        }
+
+        DB::query('COMMIT');
+
+        Hooks::dispatch('give_donation_updated', $donation);
+
+        return $donation;
+    }
+
+    /**
+     * @unreleased
+     *
+     * @param  Donation  $donation
+     * @return bool
+     * @throws Exception
+     */
+    public function delete(Donation $donation)
+    {
+        DB::query('START TRANSACTION');
+
+        Hooks::dispatch('give_donation_deleting', $donation);
+
+        try {
+            DB::table('posts')
+                ->where('id', $donation->id)
+                ->delete();
+
+            foreach ($this->getCoreDonationMetaForDatabase($donation) as $metaKey => $metaValue) {
+                DB::table('give_donationmeta')
+                    ->where('donation_id', $donation->id)
+                    ->where('meta_key', $metaKey)
+                    ->delete();
+            }
+        } catch (Exception $exception) {
+            DB::query('ROLLBACK');
+
+            Log::error('Failed deleting a donation');
```

Comment:
Seems like we could add the donation and exception for context, here? This doesn't seem very helpful as is. See other error logs in this repo as well, please.

Reply:
resolved dfe29ac4